engine/src/main/java/org/pentaho/di/trans/steps/textfileoutputlegacy/TextFileOutputLegacy.java

```diff
@@ -0,0 +1,193 @@
+import org.pentaho.di.trans.step.StepDataInterface;
+import org.pentaho.di.trans.step.StepMeta;
+import org.pentaho.di.trans.steps.textfileoutput.TextFileOutput;
+
+import java.io.IOException;
+
+/**
+ * This is deprecated version with capability run as command.
+ * @deprecated use {@link org.pentaho.di.trans.steps.textfileoutput.TextFileOutput} instead.
+ */
+@Deprecated
+public class TextFileOutputLegacy extends TextFileOutput {
+
+  public TextFileOutputLegacy( StepMeta stepMeta, StepDataInterface stepDataInterface, int copyNr, TransMeta transMeta,
+                               Trans trans ) {
+    super( stepMeta, stepDataInterface, copyNr, transMeta, trans );
+  }
+
+  @Override
+  protected boolean writeRowTo( Object[] row ) throws KettleException {
+    if ( ( (TextFileOutputLegacyMeta) meta).isFileAsCommand() ) {
+      return writeRowToCommand( row );
+    } else {
+      return super.writeRowTo( row );
+    }
+  }
+
+  private boolean writeRowToCommand( Object[] row ) throws KettleException {
+    if ( row != null ) {
+      if ( data.writer == null ) {
+        initCommandStreamWriter( environmentSubstitute( meta.getFileName() ) );
+      }
+      first = false;
+      writeRow( data.outputRowMeta, row );
+      putRow( data.outputRowMeta, row ); // in case we want it to go further...
+
+      if ( checkFeedback( getLinesOutput() ) ) {
+        logBasic( "linenr " + getLinesOutput() );
+      }
+      return true;
+    } else {
+      if ( ( data.writer == null ) && !Utils.isEmpty( meta.getEndedLine() ) ) {
+        initCommandStreamWriter( environmentSubstitute( meta.getFileName() ) );
+        initBinaryDataFields();
+      }
+      writeEndedLine();
+      closeCommand();
+      setOutputDone();
+      return false;
+    }
+  }
+
+
+  private boolean closeCommand() {
+    boolean retval;
+
+    try {
+      if ( data.writer != null ) {
+        data.writer.flush();
+
+        // If writing a ZIP or GZIP file not from a command, do not close the writer or else
+        // the closing of the ZipOutputStream below will throw an "already closed" exception.
+        // Rather than checking for compression types, it is easier to check for cmdProc != null
+        // because if that check fails, we know we will get into the ZIP/GZIP processing below.
+        if ( log.isDebug() ) {
+          logDebug( "Closing output stream" );
+        }
+        data.writer.close();
+        if ( log.isDebug() ) {
+          logDebug( "Closed output stream" );
+        }
+      }
+      data.writer = null;
+      if ( log.isDebug() ) {
+        logDebug( "Ending running external command" );
+      }
+
+      if ( data.cmdProc != null ) {
+        int procStatus = data.cmdProc.waitFor();
+        // close the streams
+        // otherwise you get "Too many open files, java.io.IOException" after a lot of iterations
+        try {
+          data.cmdProc.getErrorStream().close();
+          data.cmdProc.getOutputStream().flush();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getOutputStream*. Currently, there are execution paths that do not contain closure statements, for example, when OutputStream.flush throws an exception. Ensure the object returned by *getOutputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).